File: 155. Min Stack/solution.go
type MinStack struct {
    stack []int
    min []int
}


/** initialize your data structure here. */
func Constructor() MinStack {
    return MinStack{}
}


func (this *MinStack) Push(x int)  {
    if len(this.stack) != 0 {
        l := len(this.min)-1
        if this.min[l] > x {
            this.min = append(this.min, x)
        } else {
            this.min = append(this.min, this.min[l])
        }
    } else {
        this.min = append(this.min, x)
    }
    this.stack = append(this.stack, x)
}


func (this *MinStack) Pop()  {
    l := len(this.stack) - 1
    if l >= 0 {
        this.stack = this.stack[:l]
        this.min = this.min[:l]
    }
}


func (this *MinStack) Top() int {
    return this.stack[len(this.stack)-1]
}


func (this *MinStack) GetMin() int {
    return this.min[len(this.min)-1]
}


/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

File: 155. Min Stack/solution.py
class Node:
    def __init__(self, val, valMin):
        self.val = val
        self.min = valMin
        
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        valMin = min(self.stack[-1].min, x) if len(self.stack) > 0 else x
        node = Node(x, valMin)
        self.stack.append(node)
        # for i in range(len(self.stack)):
        #     print(self.stack[i].val, self.stack[i].min)

    def pop(self):
        """
        :rtype: void
        """
        if len(self.stack) > 0:
            self.stack.pop() 
        

    def top(self):
        """
        :rtype: int
        """
        if len(self.stack) > 0:
            return self.stack[-1].val 

    def getMin(self):
        """
        :rtype: int
        """
        if len(self.stack) > 0:
            return self.stack[-1].min 


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
